File: algorithms/02_c/graphs/bellman_ford.c
/* Bellman-Ford single-source shortest paths with negative edges. */
#include <stdio.h>
#include <stdbool.h>
#define V 4
#define E 4
#define INF 1000000

bool bellman_ford(int edges[E][3], int src, int dist[V]) {
    for (int i = 0; i < V; i++) dist[i] = INF;
    dist[src] = 0;
    for (int iter = 0; iter < V - 1; iter++)
        for (int e = 0; e < E; e++)
            if (dist[edges[e][0]] != INF &&
                dist[edges[e][0]] + edges[e][2] < dist[edges[e][1]])
                dist[edges[e][1]] = dist[edges[e][0]] + edges[e][2];
    for (int e = 0; e < E; e++)
        if (dist[edges[e][0]] != INF &&
            dist[edges[e][0]] + edges[e][2] < dist[edges[e][1]])
            return false;
    return true;
}

int main(void) {
    int edges[E][3] = {{0, 1, 4}, {0, 2, 5}, {1, 2, -3}, {2, 3, 2}};
    int dist[V];
    if (!bellman_ford(edges, 0, dist)) return 1;
    for (int i = 0; i < V; i++) printf("%d ", dist[i]);
    printf("\n");
    return 0;
}

File: algorithms/11_golang/graphs/bellman_ford.go
// Bellman-Ford shortest paths with negative edge support.
package main

import "fmt"

const inf = 1 << 60

type edge struct {
	from, to, weight int
}

func bellmanFord(n int, edges []edge, source int) ([]int, bool) {
	dist := make([]int, n)
	for i := range dist {
		dist[i] = inf
	}
	dist[source] = 0
	for iter := 0; iter < n-1; iter++ {
		for _, e := range edges {
			if dist[e.from] != inf && dist[e.from]+e.weight < dist[e.to] {
				dist[e.to] = dist[e.from] + e.weight
			}
		}
	}
	for _, e := range edges {
		if dist[e.from] != inf && dist[e.from]+e.weight < dist[e.to] {
			return nil, false
		}
	}
	return dist, true
}

func main() {
	edges := []edge{{0, 1, 4}, {0, 2, 5}, {1, 2, -3}, {2, 3, 2}}
	dist, ok := bellmanFord(4, edges, 0)
	if !ok {
		panic("unexpected negative cycle")
	}
	fmt.Println(dist)
}
